### PR TITLE
webui: separate search results by job ids

### PR DESCRIPTION
**Pull request by junhaoliao**
# References
Kirk reported that race conditions in search results and/or metadata were discovered when the server was undergoing high load.

# Description
1. Create a single search metadata collection for all and store each job's results metadata per document.
2. Instead of storing search results per session, store those per job with the collection name being `results_${job_id}`.
3. In the webui-query-handler, pass `job_id` as part of the state reporting back to the meteor server, which is later propagated to the meteor client side.
4. For each search / timeline update, pull results from `results_${job_id}` and metadata from `results-metadata`'s document `meta-job-${job_id}`.

# Validation performed
Manual integration test. Steps include:
1. Rebuild `clp-package`.
2. Start `clp-package`.
3. Session A: In the Chromium-based Microsoft Edge browser, visit `http://localhost:4000` and start a search query.
4. Session B: Open a "New InPrivate Window" in Microsoft Edge and visit `http://localhost:4000`, effectively emulating another user visiting the same site. Start a search query.
5. Keep both Session A and B open, adjust pipeline strings (search keywords) in both Session A and B, and observe search results showing up independently in each session.
6. Keep both Session A and B open, adjust the timeline in both Session A and B, and observe updated search results showing up independently in each session.

### CODE REVIEW

File: components/webui-query-handler/webui_query_handler/main.py [hunk 1]
```diff
@@ -534,10 +533,12 @@ async def cancel_metadata_task(task, pending_tasks, query_done_event):
         query_done_event.clear()
 
 
+# FIXME: only clear specific job in meta
```
Comment (LinZhihao-723) [inline components/webui-query-handler/webui_query_handler/main.py:536]: Can we move this inline comment to the function body and mark it as `TODO`?

File: components/webui/imports/api/search/server/publications.js [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
-        };
-        MyCollections[sessionId][Meteor.settings.public.SearchResultsCollectionName] =
-            new Mongo.Collection(`${Meteor.settings.public.SearchResultsCollectionName}_${sessionId}`);
-        MyCollections[sessionId][Meteor.settings.public.SearchResultsMetadataCollectionName] =
-            new Mongo.Collection(`${Meteor.settings.public.SearchResultsMetadataCollectionName}_${sessionId}`);
+const createCollectionsIfNotExist = (sessionId, jobID) => {
+    const collectionName = `${Meteor.settings.public.SearchResultsCollectionName}_${jobID}`;
+
+    if ((MyCollections[sessionId] !== undefined) &&
+        (MyCollections[sessionId][Meteor.settings.public.SearchResultsCollectionName]._name === collectionName)) {
```
Comment (LinZhihao-723) [inline components/webui/imports/api/search/server/publications.js:11]:
```suggestion
    if (undefined !== MyCollections[sessionId] &&
        collectionName === MyCollections[sessionId][Meteor.settings.public.SearchResultsCollectionName]._name) {
```
If possible, we should put rvalue on the rhs for comparison

File: components/webui/imports/ui/SearchView/SearchView.jsx [hunk 3]
```diff
@@ -109,33 +117,43 @@ const SearchView = () => {
             findOptions["sort"] = sort;
         }
 
+        if ((!serverState) || !(serverState.jobID)) {
```
Comment (LinZhihao-723) [inline components/webui/imports/ui/SearchView/SearchView.jsx:120]:
```suggestion
        if (null === serverState || undefined === serverState.jobID) {
```
I am not sure if it is possible for `serverState.jobID` to be `null` as well

File: components/webui/imports/ui/SearchView/SearchView.jsx [hunk 4]
```diff
@@ -109,33 +117,43 @@ const SearchView = () => {
...
-            if (numSearchResultsDoc) {
-                numSearchResultsOnServer = numSearchResultsDoc["all"];
-            }
-
-            const timelineDoc = MyCollections[Meteor.settings.public.SearchResultsMetadataCollectionName].findOne({_id: "timeline"});
-            if (timelineDoc) {
-                timeline = timelineDoc;
+            // FIXME: add job id
+            const metaDoc = SearchResultsMetadataCollection.findOne({_id: `meta-job-${serverState.jobID}`});
+            if (metaDoc) {
```
Comment (LinZhihao-723) [inline components/webui/imports/ui/SearchView/SearchView.jsx:139]:
```suggestion
            if (null !== metaDoc) {
```

File: components/webui/imports/ui/SearchView/SearchView.jsx [hunk 5]
```diff
@@ -604,7 +624,7 @@ const SearchResultsTable = ({
                     }
 
                     let result = searchResult[columnName];
-                    if (typeof(result) === "object") {
+                    if (typeof (result) === "object") {
```
Comment (LinZhihao-723) [inline components/webui/imports/ui/SearchView/SearchView.jsx:627]: Why do we need the space here?

File: components/webui/imports/api/search/server/query_handler_mediator.js [hunk 6]
```diff
@@ -88,6 +88,7 @@ class WebUiQueryHandlerWebsocket {
             case ServerMessageType.QUERY_STARTED:
                 if (SearchState.WAITING_FOR_QUERY_TO_START === currentServerStateList[this.__sessionId].state) {
                     currentServerStateList[this.__sessionId].state = SearchState.QUERY_IN_PROGRESS;
+                    currentServerStateList[this.__sessionId].jobID = message['value']['jobID']
```
Comment (LinZhihao-723) [inline components/webui/imports/api/search/server/query_handler_mediator.js:91]:
```suggestion
                    currentServerStateList[this.__sessionId].jobID = message['value']['jobID'];
```

File: components/webui/imports/ui/SearchView/SearchView.jsx [hunk 7]
```diff
@@ -109,33 +117,43 @@ const SearchView = () => {
...
         if (isReady) {
-            const numSearchResultsDoc = MyCollections[Meteor.settings.public.SearchResultsMetadataCollectionName].findOne({_id: "num_search_results"});
-            if (numSearchResultsDoc) {
-                numSearchResultsOnServer = numSearchResultsDoc["all"];
-            }
-
-            const timelineDoc = MyCollections[Meteor.settings.public.SearchResultsMetadataCollectionName].findOne({_id: "timeline"});
-            if (timelineDoc) {
-                timeline = timelineDoc;
+            // FIXME: add job id
```
Comment (LinZhihao-723) [inline components/webui/imports/ui/SearchView/SearchView.jsx:137]: Can we use `TODO` instead?

**Review by LinZhihao-723 — COMMENTED** [review]
Added some comments on the coding style. I will leave the core functionality review to Kirk

**Comment by junhaoliao** [comment]
Closing this as refactoring changes has been proposed and being PR'ed at #250 .

**Pull request closed without merging**